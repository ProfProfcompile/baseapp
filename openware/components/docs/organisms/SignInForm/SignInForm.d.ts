import * as React from 'react';
interface SignInFormValues {
    email: string;
    password: string;
}
interface SignInFormProps {
    /**
     * Submit error message
     */
    errorMessage?: string;
    /**
     * Enables loader when it is true
     */
    isLoading?: boolean;
    /**
     * Form title (e.g. 'Log in')
     * @default 'Sign-in'
     */
    title?: string;
    /**
     * Function which will be called on forgot password
     */
    onForgotPassword: (email?: string) => void;
    /**
     * Function which will be called on sign up
     */
    onSignUp: () => void;
    /**
     * Function which will be called on sign in
     */
    onSignIn: (data: SignInFormValues) => void;
    /**
     *  By default class name 'cr-sign-in-form'
     *  This property gives an additional class name
     */
    className?: string;
    /**
     * Customer's logo
     */
    image?: string;
}
export interface SignInFormState {
    email: string;
    emailError: string;
    password: string;
    passwordError: string;
}
declare class SignInForm extends React.Component<SignInFormProps, SignInFormState> {
    constructor(props: SignInFormProps);
    render(): JSX.Element;
    private handleSubmitForm;
    private isValidForm;
    private validateForm;
    private handleInput;
    private handleClick;
}
export { SignInForm, SignInFormProps, SignInFormValues, };
