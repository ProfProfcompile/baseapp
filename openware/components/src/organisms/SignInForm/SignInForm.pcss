.cr-sign-in-form {
  padding: calc(var(--gap) * 2) calc(var(--gap) * 8);
  background: var(--sign-in-background-color);
  max-width: 500px;

  & > * {
    background: var(--transparent);
  }

  &__error-message {
    color: var(--color-red);
    font-size: calc(var(--gap) * 2.5);
    flex: 1;
    display: flex;
    background: var(--transparent);
    justify-content: center;
    align-items: center;
    min-height: calc(var(--gap) * 6);
  }

  &__group {
    margin-bottom: 20px;
    position: relative;
  }

  &__error {
    background: var(--transparent);
    margin-top: 5px;
    color: var(--color-red);
  }

  &__label {
    background: var(--sign-in-background-color);
    color: var(--sign-in-label-color);
  }

  &__image {
    display: block;
    width: 50%;
    height: 100px;
    margin: 0 auto 20px auto;
  }

  &__title {
    color: var(--sign-in-label-color);
    margin: calc(var(--gap) * 3) 0 calc(var(--gap) * 8) 0;
    font-size: 18px;
    text-align: center;
  }

  &__button {
    background: var(--button-primary);
    height: 50px;
    font-size: 18px;
    font-weight: normal;
    margin-left: 0;
  }

  &__loader {
    background: var(--transparent);
    min-height: calc(var(--gap) * 6);
  }

  &__button-wrapper {
    min-height: calc(var(--gap) * 28);
    display: flex;
    flex-direction: column;
    justify-content: space-between;
  }

  &__button-wrapper--empty {
    justify-content: flex-end;
  }

  &__input {
    width: 100%;

    input {
      color: var(--sign-in-input-color);
      width: 100%;
      height: 30px;
      font-size: 12px;
      text-align: left;
      border-bottom: 1px solid var(--sign-in-input-border-color);
    }
  }

  &__footer {
    display: flex;
    justify-content: center;
    align-items: center;

    & > * {
      background: var(--transparent);
    }

    &-create {
      color: var(--sign-in-label-color);
    }

    &-signup {
      cursor: pointer;
      padding-left: calc(var(--gap) * 2);
    }
  }

  &__checkbox {
    word-wrap: break-word;
    background: var(--transparent);

    .cr-checkbox__checkitem {
      border: 1px solid var(--sign-up-checkbox-mark-color);
      background: var(--transparent);

      &::after {
        border-color: var(--sign-up-checkbox-mark-color);
      }
    }

    input:checked ~ .cr-checkbox__checkitem {
      background-color: var(--transparent);
    }

    .cr-checkbox__label {
      white-space: normal;
      padding-top: 0;
      font-size: 12px;
      background: var(--transparent);
      color: var(--sign-up-checkbox-color);
    }
  }

  &__forgot {
    color: var(--sign-up-checkbox-color);
    cursor: pointer;
    position: absolute;
    top: calc(var(--gap) * 2);
    font-size: calc(var(--gap) * 2.5);
    background: var(--transparent);
    right: 0;
  }
}
