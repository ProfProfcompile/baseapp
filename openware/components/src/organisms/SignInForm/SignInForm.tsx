/* tslint:disable:jsx-no-lambda*/
import cr from 'classnames';
import * as React from 'react';

import {Button, Input, Loader} from '../../atoms';
import {
    EMAIL_REGEX,
    ERROR_EMPTY_PASSWORD,
    ERROR_INVALID_EMAIL,
} from '../../constants';

interface SignInFormValues {
    email: string;
    password: string;
}

interface SignInFormProps {
    /**
     * Submit error message
     */
    errorMessage?: string;
    /**
     * Enables loader when it is true
     */
    isLoading?: boolean;
    /**
     * Form title (e.g. 'Log in')
     * @default 'Sign-in'
     */
    title?: string;
    /**
     * Function which will be called on forgot password
     */
    onForgotPassword: (email?: string) => void;
    /**
     * Function which will be called on sign up
     */
    onSignUp: () => void;
    /**
     * Function which will be called on sign in
     */
    onSignIn: (data: SignInFormValues) => void;
    /**
     *  By default class name 'cr-sign-in-form'
     *  This property gives an additional class name
     */
    className?: string;
    /**
     * Customer's logo
     */
    image?: string;
}

export interface SignInFormState {
    email: string;
    emailError: string;
    password: string;
    passwordError: string;
}

class SignInForm extends React.Component<SignInFormProps, SignInFormState> {
    constructor(props: SignInFormProps) {
        super(props);
        this.state = {
            email: '',
            emailError: '',
            password: '',
            passwordError: '',
        };
    }

    public render() {
        const {
            email,
            emailError,
            password,
            passwordError,
        } = this.state;
        const {
            errorMessage,
            isLoading,
            onSignUp,
            onForgotPassword,
            image,
        } = this.props;
        const buttonWrapperClass = cr('cr-sign-in-form__button-wrapper', {
            'cr-sign-in-form__button-wrapper--empty': !errorMessage,
        });
        return (
            <form>
                <div className="cr-sign-in-form">
                    {image ? (
                        <h1 className="cr-sign-in-form__title">
                            <img className="cr-sign-in-form__image" src={image} alt="logo"/>
                            Sign-in
                        </h1>
                    ) : (
                        <h1 className="cr-sign-in-form__title" style={{ marginTop: 119 }}>
                            Sign-in
                        </h1>
                        )}

                    <div className="cr-sign-in-form__group">
                        <label className="cr-sign-in-form__label">Email</label>
                        <Input
                            type={'email'}
                            value={email}
                            className={'cr-sign-in-form__input'}
                            onChangeValue={value => this.handleInput(value, 'email')}
                        />
                        {emailError && <div className={'cr-sign-in-form__error'}>{emailError}</div>}
                    </div>
                    <div className="cr-sign-in-form__group">
                        <label className="cr-sign-in-form__label">Password</label>
                        <Input
                            type={'password'}
                            value={password}
                            className={'cr-sign-in-form__input'}
                            onChangeValue={value => this.handleInput(value, 'password')}
                        />
                        <div
                            className={'cr-sign-in-form__forgot'}
                            onClick={() => onForgotPassword(email)}
                        >
                            Forgot?
                        </div>
                        {passwordError && <div className={'cr-sign-in-form__error'}>{passwordError}</div>}
                    </div>
                    <div className={buttonWrapperClass}>
                        <div className="cr-sign-in-form__error-message">{errorMessage || null}</div>
                        <div className="cr-sign-in-form__loader">{isLoading ? <Loader/> : null}</div>
                        <Button
                            label={isLoading ? 'Loading...' : 'Sign in'}
                            type="submit"
                            className={'cr-sign-in-form__button'}
                            disabled={isLoading}
                            onClick={this.handleClick}
                        />
                    </div>
                    <div className={'cr-sign-in-form__footer'}>
                        <p className={'cr-sign-in-form__footer-create'}>Create an account?</p>
                        <a
                            className={'cr-sign-in-form__footer-signup'}
                            onClick={onSignUp}
                        >
                            Sign up
                        </a>
                    </div>
                </div>
            </form>
        );
    }

    private handleSubmitForm() {
        const {
            email,
            password,
        } = this.state;

        this.setState({
            emailError: '',
            passwordError: '',
        }, () => {
            this.props.onSignIn({email, password});
        });
    }

    private isValidForm() {
        const {email, password} = this.state;
        const isEmailValid = email.match(EMAIL_REGEX);

        return email && isEmailValid && password;
    }

    private validateForm() {
        const {email,password} = this.state;
        const isEmailValid = email.match(EMAIL_REGEX);

        if (!isEmailValid) {
            this.setState({
                emailError: ERROR_INVALID_EMAIL,
                passwordError: '',
            });
            return;
        }
        if (!password) {
            this.setState({
                emailError: '',
                passwordError: ERROR_EMPTY_PASSWORD,
            });
            return;
        }
    }

    private handleInput = (value: string, type: string) => {
        switch (type) {
            case 'email':
                this.setState({
                    email: value,
                });
                break;
            case 'password':
                this.setState({
                    password: value,
                });
                break;
            default:
                break;
        }
    };

    private handleClick = (label?: string, e?: React.FormEvent<HTMLInputElement>) => {
        if (e) {
            e.preventDefault();
        }
        if (!this.isValidForm()) {
            this.validateForm();
        } else {
            this.handleSubmitForm();
        }
    };
}


export {
    SignInForm,
    SignInFormProps,
    SignInFormValues,
};
