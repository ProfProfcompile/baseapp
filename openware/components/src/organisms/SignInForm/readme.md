Sign in form example:
```js
<SignInForm
    onForgotPassword={(email) => console.log(email)}
    onSignIn={(data) => console.log(data)}
    onSignUp={() => console.log('Sign up page')}
/>
```
