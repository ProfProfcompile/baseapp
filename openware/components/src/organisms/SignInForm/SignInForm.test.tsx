import { shallow } from 'enzyme';
import * as React from 'react';
import { SinonSpy, spy } from 'sinon';
import { SignInForm, SignInFormProps, SignInFormState } from './SignInForm';

const defaults: SignInFormProps = {
    onForgotPassword: jest.fn(),
    onSignIn: jest.fn(),
    onSignUp: jest.fn(),
};

const setup = (props: Partial<SignInFormProps> = {}) =>
    shallow(<SignInForm {...{ ...defaults, ...props }} />);

describe('Sign in form', () => {
    let wrapper = setup();

    beforeEach(() => {
        wrapper = setup();
    });

    it('should render', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should have correct className', () => {
        expect(wrapper.hasClass('cr-sign-in-form'));
    });

    it('should handle forgot onClick', () => {
        const onClick = spy();
        wrapper = setup({ onForgotPassword: onClick });
        wrapper.setState({ email: 'potato@heliostech.fr' });
        wrapper.find('.cr-sign-in-form__forgot').simulate('click');
        expect((onClick as SinonSpy).calledOnceWith('potato@heliostech.fr')).toBeTruthy();
    });

    it('should handle SignIn onClick', () => {
        const onClick = spy();
        wrapper = setup({ onSignIn: onClick });
        wrapper.setState({ email: 'potato@heliostech.fr', password: 'Potato3322' });
        wrapper.find('.cr-sign-in-form__button').simulate('click');
        expect((onClick as SinonSpy).calledOnceWith({email: 'potato@heliostech.fr', password: 'Potato3322'})).toBeTruthy();
    });

    it('should handle SignUp onClick', () => {
        const onClick = spy();
        wrapper = setup({ onSignUp: onClick });
        wrapper.find('.cr-sign-in-form__footer-signup').simulate('click');
        expect((onClick as SinonSpy).calledOnce).toBeTruthy();
    });

    it('should handle input valid email value', () => {
        wrapper.props().children.props.children[1].props.children[1].props.onChangeValue('potato@heliostech.fr');
        // @ts-ignore
        const state: SignInFormState = wrapper.state();
        expect(state.email).toEqual('potato@heliostech.fr');
    });

    it('should handle input valid password value', () => {
        // tslint:disable-next-line:no-magic-numbers
        wrapper.props().children.props.children[2].props.children[1].props.onChangeValue('Potato3322');
        // @ts-ignore
        const state: SignInFormState = wrapper.state();
        expect(state.password).toEqual('Potato3322');
    });

    it('should handle input invalid email value', () => {
        wrapper.setState({ email: 'potato', password: 'Potato3322' });
        wrapper.find('.cr-sign-in-form__button').simulate('click');
        // @ts-ignore
        const state: SignInFormState = wrapper.state();
        expect(state.emailError).toEqual('Email is invalid');
    });

    it('should handle input empty password value', () => {
        wrapper.setState({ email: 'potato@heliostech.fr', password: '' });
        wrapper.find('.cr-sign-in-form__button').simulate('click');
        // @ts-ignore
        const state: SignInFormState = wrapper.state();
        expect(state.passwordError).toEqual('Password is invalid');
    });

    it('should render Loader', () => {
        // tslint:disable-next-line:no-magic-numbers
        expect(wrapper.props().children.props.children[3].props.children[1].props.children).toBeNull();
        // tslint:disable-next-line:no-magic-numbers
        expect(wrapper.props().children.props.children[3].props.children[2].props.label).toEqual('Sign in');
        wrapper = setup({ isLoading: true });
        // tslint:disable-next-line:no-magic-numbers
        expect(wrapper.props().children.props.children[3].props.children[1].props.children).not.toBeNull();
        // tslint:disable-next-line:no-magic-numbers
        expect(wrapper.props().children.props.children[3].props.children[2].props.label).toEqual('Loading...');
    });

    it('should render image', () => {
       wrapper = setup({ image: 'qwer' });
       expect(wrapper.props().children.props.children[0].props.children[0].props.src).toEqual('qwer');
    });
});
