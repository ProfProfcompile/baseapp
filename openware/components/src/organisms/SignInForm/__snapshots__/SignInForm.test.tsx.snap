// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sign in form should render 1`] = `
<form>
  <div
    className="cr-sign-in-form"
  >
    <h1
      className="cr-sign-in-form__title"
      style={
        Object {
          "marginTop": 119,
        }
      }
    >
      Sign-in
    </h1>
    <div
      className="cr-sign-in-form__group"
    >
      <label
        className="cr-sign-in-form__label"
      >
        Email
      </label>
      <Input
        className="cr-sign-in-form__input"
        onChangeValue={[Function]}
        type="email"
        value=""
      />
    </div>
    <div
      className="cr-sign-in-form__group"
    >
      <label
        className="cr-sign-in-form__label"
      >
        Password
      </label>
      <Input
        className="cr-sign-in-form__input"
        onChangeValue={[Function]}
        type="password"
        value=""
      />
      <div
        className="cr-sign-in-form__forgot"
        onClick={[Function]}
      >
        Forgot?
      </div>
    </div>
    <div
      className="cr-sign-in-form__button-wrapper cr-sign-in-form__button-wrapper--empty"
    >
      <div
        className="cr-sign-in-form__error-message"
      />
      <div
        className="cr-sign-in-form__loader"
      />
      <Button
        className="cr-sign-in-form__button"
        label="Sign in"
        onClick={[Function]}
        type="submit"
      />
    </div>
    <div
      className="cr-sign-in-form__footer"
    >
      <p
        className="cr-sign-in-form__footer-create"
      >
        Create an account?
      </p>
      <a
        className="cr-sign-in-form__footer-signup"
        onClick={[MockFunction]}
      >
        Sign up
      </a>
    </div>
  </div>
</form>
`;
